tokio-boring/src/async_callbacks.rs

```diff
@@ -0,0 +1,148 @@
+use boring::ex_data::Index;
+use boring::ssl::{self, ClientHello, Ssl, SslContextBuilder};
+use once_cell::sync::Lazy;
+use std::future::Future;
+use std::pin::Pin;
+use std::task::{ready, Context, Poll};
+
+type BoxSelectCertFuture = Pin<
+    Box<
+        dyn Future<Output = Result<BoxSelectCertFinish, AsyncSelectCertError>>
+            + Send
+            + Sync
+            + 'static,
+    >,
+>;
+
+type BoxSelectCertFinish =
+    Box<dyn FnOnce(ClientHello<'_>) -> Result<(), AsyncSelectCertError> + 'static>;
+
+pub(crate) static TASK_CONTEXT_INDEX: Lazy<Index<Ssl, usize>> =
+    Lazy::new(|| Ssl::new_ex_index().unwrap());
+pub(crate) static SELECT_CERT_FUTURE_INDEX: Lazy<Index<Ssl, BoxSelectCertFuture>> =
+    Lazy::new(|| Ssl::new_ex_index().unwrap());
+
+/// Extensions to [`SslContextBuilder`].
+///
+/// This trait provides additional methods to use async callbacks with boring.
+pub trait SslContextBuilderExt: private::Sealed {
```

Comment:
At first I thought I would just expose `async_select_certificate_callback` and `polling_select_certificate_callback` but using an extension trait makes usage more convenient.